### PR TITLE
Updated source code for Chapter 3.

### PR DESCRIPTION
**Pull request by MarkMichaelis**

### CODE REVIEW

File: src/Shared/CompilerAssert.cs [hunk 1]
```diff
@@ -68,52 +64,60 @@ public bool Equals(CompileError rhs)
...
-            }
+        }
 
-            async static public Task<CompileError[]> ExpectErrorsAsync(
+        async static public Task<CompileError[]> ExpectErrorsAsync(
             string sourceCode, params CompileError[] diagnostics)
         {
+            CompileError[] actualCompileErrors = await CompileAsync(sourceCode);
+
+            if (diagnostics is { } && diagnostics.Length > 0)
```
Comment (Keboo) [inline src/Shared/CompilerAssert.cs]:
```suggestion
            if (diagnostics?.Length > 0)
```

File: src/Shared/CompilerAssert.cs [hunk 2]
```diff
@@ -68,52 +64,60 @@ public bool Equals(CompileError rhs)
...
+                        $"The expected Ids do not match for item {i}: " +
+                        $"{diagnostics[i].Id}: {diagnostics[i].Message} <> {actualCompileErrors[i].Id}: {actualCompileErrors[i].Message}");
+                }
+                return actualCompileErrors;
+            }
+
+            Assert.Fail("The expected compilation errors did not occur.");
+            return null!;  // This code will never execute due to the Fail line above.
+        }
+        async static public Task<CompileError[]> CompileAsync(
```
Comment (Keboo) [inline src/Shared/CompilerAssert.cs]:
```suggestion
        public static async Task<CompileError[]> CompileAsync(
```

File: src/Shared/CompilerAssert.cs [hunk 3]
```diff
@@ -41,8 +37,8 @@ public override string ToString()
         // override object.Equals
         public override bool Equals(object? obj)
         {
-            return (obj is CompileError)
-              && Equals((CompileError)obj);
+            return (obj is CompileError error)
+              && Equals(error);
```
Comment (Keboo) [inline src/Shared/CompilerAssert.cs:41]: Making this a record strut would eliminate most of the code for `CompileError`

File: src/Shared/Tests/Parallel.runsettings [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Keboo) [inline src/Shared/Tests/Parallel.runsettings:1]: Are we configuring something to use this file?

**Review by Keboo — APPROVED** [review]

File: src/Shared/CompilerAssert.cs (hunk 3 above)
Comment (MarkMichaelis) [reply src/Shared/CompilerAssert.cs:41]: Agreed, perhaps even a record struct. I chose not to take on the distraction in spite of the temptation. :)

File: src/Shared/Tests/Parallel.runsettings (hunk 4 above)
Comment (MarkMichaelis) [reply src/Shared/Tests/Parallel.runsettings:1]: I'd like to have it used by default when running tests but my understanding is that it requires a reference in the command line (or perhaps an environment variable works?).  Suggestions welcome.

File: src/Shared/CompilerAssert.cs (hunk 3 above)
Comment (MarkMichaelis) [reply src/Shared/CompilerAssert.cs:41]: Agreed.... there wasn't enough value in the change so I avoided the distraction. See #393.

File: src/Shared/Tests/Parallel.runsettings (hunk 4 above)
Comment (Keboo) [reply src/Shared/Tests/Parallel.runsettings:1]: I think MSBuild properties are you best bet. https://learn.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file?view=vs-2022#set-a-build-property

**Pull request merged**